R/teal_lockfile.R

```diff
@@ -0,0 +1,119 @@
+#' - Automatically computed:
+#'     - **Working directory lockfile**: If `options("teal.renv.lockfile")` is empty, `teal` will, by default, create an
+#'      `implicit` type lockfile that uses `renv::dependencies()` to detect all R packages in the current project's
+#'      working directory.
+#'     - **`DESCRIPTION`-based lockfile**: To generate a lockfile based on a `DESCRIPTION` file in your working
+#'     directory, set `renv::settings$snapshot.type("explicit")`. The naming convention for `type` follows
+#'     `renv::snapshot()`. For the `"explicit"` type, refer to `renv::settings$package.dependency.fields()` for the
+#'     `DESCRIPTION` fields included in the lockfile.
+#'     - **Custom files-based lockfile**: To specify custom files as the basis for the lockfile, set
+#'     `renv::settings$snapshot.type("custom")` and configure the `renv.snapshot.filter` option.
+#'
+#' @note
+#' This function computes the lockfile as a `promises::future_promise` promise, executing the process on a separate
+#' worker. If `future::plan()` is set to something other than `future::sequential`, it reuses the existing parallel
+#' backend. Otherwise, it sets `future::multisession` as the parallel plan with two workers. The `shiny::ExtendedTask()`
+#' is then used to run asynchronous computations. If the `teal` app started with the `future::sequential` plan, it is
+#' reset once the task is completed.
+#'
+#' @section lockfile usage:
+#' After creating the lockfile, you can restore the application environment using `renv::restore()`.
+#'
+#' @seealso [renv::snapshot()], [renv::restore()].
+#'
+#' @return Nothing. This function is executed for its side effect of creating a lockfile used in the `teal` application.
+#'
+#' @keywords internal
+teal_lockfile <- function() {
+  # If user has setup the file, there is no need to compute a new one.
+  user_lockfile <- getOption("teal.renv.lockfile", "")
+  if (!identical(user_lockfile, "")) {
+    if (file.exists(user_lockfile)) {
+      file.copy(user_lockfile, "teal_app.lock")
+      return(invisible(NULL))
+    } else {
+      stop("lockfile provided through options('teal.renv.lockfile') does not exist.")
+    }
+  }
+
+  if (!(is_in_test() || is_r_cmd_check())) {
+    old_plan <- future::plan()
+    # If there is already a parallel backend, reuse it.
+    if (inherits(old_plan, "sequential")) {
+      future::plan(future::multisession, workers = 2)
+    }
+
+    lockfile_task <- ExtendedTask$new(create_renv_lockfile)
+    lockfile_task$invoke(inherits(old_plan, "sequential"))
+    logger::log_trace("lockfile creation invoked.")
+  }
+}
+
+create_renv_lockfile <- function(close) {
+  checkmate::assert_flag(close)
+  promise <- promises::future_promise({
+    # Below we can not use a file created in tempdir() directory.
+    # If a file is created in tempdir(), it gets deleted on 'then(onFulfilled' part.
+    lockfile_path <- "teal_app.lock"
+    shiny::onStop(function() file.remove(lockfile_path))
```

Comment:
just because of this I think we can't do `file.move()` but rather `file.copy()` instead.
We are going to loose user provided file on the very first app exit

Reply:
@pawelru we have `file.copy` in `teal_lockfile`.
So if user provides his/her own file (let's say called `x.lock`), it is copied with a new name `teal_app.lock`. Then `teal_app.lock` is deleted on app exit, but the initial file `x.lock` is not deleted.

Reply:
ok ok
I don't know why I was thinking athat we have file move. This is good